[DataGridPro] Tree data row reordering support (#19401)
mui/mui-x: docs/data/data-grid/tree-data/tree-data.md

:::

> ## Drag-and-drop tree data reordering

Reviewer: Optional - to make it aligned with the subsections above.
```suggestion
## Tree data row reordering
```

Author: I wanted to align it with https://mui.com/x/react-data-grid/row-grouping/#drag-and-drop-group-reordering, would be curious how'd @mapache-salvaje approach this.

With row reordering, users can reorder tree data or move rows from one group to another.

> To enable this feature with row grouping, pass the `rowReordering` prop to the Data Grid component:

Reviewer: ```suggestion
To enable this feature with tree data, pass the `rowReordering` prop to the Data Grid component:
```

```tsx
<DataGridPro
  columns={columns}
  rows={rows}
  treeData
  getTreeDataPath={getTreeDataPath}
  setTreeDataPath={setTreeDataPath}
  rowReordering
/>
```

> :::warning

Reviewer 2: ```suggestion
```

This warning can be removed. To me, those methods are required to make the feature work, it's not a warning.

> In order for the cross parent operations to work where there will be a change in the path, you need to pass the `setTreeDataPath()` prop, that works reverse to how [`getTreeDataPath()`](/x/api/data-grid/data-grid-pro/#data-grid-pro-prop-getTreeDataPath) works.

Reviewer: ```suggestion
Row reordering can change the path of the row.
Pass the `setTreeDataPath()` prop to revert the operation done by [`getTreeDataPath()`](/x/api/data-grid/data-grid-pro/#data-grid-pro-prop-getTreeDataPath) while building the tree.
```

:::

{{"demo": "TreeDataReordering.js", "bg": "inline", "defaultCodeOpen": false}}

> ### Sync row data with reordered data

Reviewer: ```suggestion
### Reordering controlled rows
```

Reviewer 3: Why? I feel like this section should be called "Reordering persistance" instead

Author: I also think that using the word "controlled" might cause, as "rows" is a semi-controlled'ish prop which is only controlled one way (from outside to the Data Grid).

> should be called "Reordering persistance" instead

Keeping in mind explanation in https://github.com/mui/mui-x/pull/19401/files#r2395786742 I think this is closer to the users' expectation. Will use it.

> If you want to update the external row data, for example to persist it in the local storage, you can use the `onRowOrderChange()` callback and the Data Grid selectors to get the reordered data and sync with the external data.

Reviewer: ```suggestion
If you want to update the external row data, for example, to persist it in the local storage, you can use the `onRowOrderChange()` callback and the Data Grid selectors to get the new row order and sync with the external data.
```

```tsx

> <DataGridPro onRowOrderChange={handleRowOrderChange} />

Reviewer 3: Shouldn't persistance be implemented with the `processRowUpdate` prop?
Same as with editing and row reordering with row grouping?
From what I see in the code, this is already implemented after `setTreeDataPath` returns an updated row.

Reviewer 3: So for row reordering persistence to work with nested data, both `processRowUpdate` and `onRowOrderChange` should be used, correct?

Reviewer 3: Thanks for the details!

> 3. Proper backend data persistence with order tracking

I think this is the scenario we should follow in our persistence demo.
Initially, I thought that the need to use both `processRowUpdate` and `onRowOrderChange` for persistence is weird. But on second thought, that's exactly what you'd do in case of a flat rows structure if you want to support Editing and Row reordering. The difference is that for nested rows, moving the row to a different parent === editing operation, so you have to use Editing API to persist such an action 👍🏻

In our docs we should explain that moving a row from one parent to another edits a row under the hood, and this is why the `processRowUpdate` is necessary for persistance.

Author: Updated the demo with a local storage based store that can mimic an actual backend which tracks the order of nested rows in a tree structure: https://deploy-preview-19401--material-ui-x.netlify.app/x/react-data-grid/tree-data/#reordering-persistance

Let me know if it makes sense now?

```tsx
<DataGridPro
  columns={columns}
  rows={rows}
  treeData
  getTreeDataPath={getTreeDataPath}
  setTreeDataPath={setTreeDataPath}
  rowReordering
/>
```

{{"demo": "TreeDataReordering.js", "bg": "inline", "defaultCodeOpen": false}}

### Reordering persistance

To sync the updated row order with an external store, depending on how you manage the order of rows in the external store, you can use either [`processRowUpdate()`](/x/api/data-grid/data-grid-pro/#data-grid-pro-prop-processRowUpdate) or the [`onRowOrderChange()`](/x/api/data-grid/data-grid-pro/#data-grid-pro-prop-onRowOrderChange) callback, or both.

The `processRowUpdate()` callback is triggered whenever a row is updated, so when doing a cross-parent reorder operation, the `path` value gets updated based on the new position of the row(s) which result in one or more `processRowUpdate()` calls. If the `path` value is part of the row data, you can use the `processRowUpdate()` callback to update the `path` value.

> After the reorder operation is successfully completed, the `onRowOrderChange()` callback is triggered, which contains information about the new row re-ordering of the format [`GridRowOrderChangeParams`](/x/api/data-grid/grid-row-order-change-params/).

Reviewer 4: ```suggestion
After the reorder operation is successfully completed, the `onRowOrderChange()` callback is triggered, which contains information about the new row reordering of the format [`GridRowOrderChangeParams`](/x/api/data-grid/grid-row-order-change-params/).
```

```tsx
<DataGridPro
  // Capture row updates, can be multiple in one reorder operation, only triggered

>   // when one or more row updates (i.e. a cross parent reorder operation) are performed

Reviewer 4: ```suggestion
  // when one or more row updates (such as a cross-parent reorder operation) are performed
```

  // when one or more row updates (such as a cross-parent reorder operation) are performed
  processRowUpdate={processRowUpdate}
  // Capture row reordering operations, always triggered once
  // after the reorder operation is successfully completed
  onRowOrderChange={handleRowOrderChange}
/>
```

The demo below uses a custom data store bound using `useSyncExternalStore()` to persist the row data in local storage.

> It maintains a row tree structure which is used to recompute path values after a reorder operation is performed, so `processRowUpdate()` is not needed—`onRowOrderChange()` alone is sufficient for syncing the updated row order with the external store.

Reviewer 3: Nice demo with files BTW!

:::success
**Tip**: After some reorders, try refreshing the demo/page to see the persisted row order.
:::

{{"demo": "TreeDataSyncRowData.js", "bg": "inline", "defaultCodeOpen": false}}

:::info

> The demo above uses the `isValidRowReorder()` prop to disable moving nodes "over" files (or leaf nodes) as it would convert them into parent groups.

Reviewer 4: ```suggestion
The demo above uses the `isValidRowReorder()` prop to disable moving nodes "over" files (or leaf nodes) because this would convert them into parent groups.
```
